File: spec/game/runtime_settings_spec.lua
local runtime_settings = require("game.runtime_settings")
local combat_feedback = require("game.presentation.combat_feedback")
local settings = require("game.settings")
local t = require("spec.support.runner")

t.describe("runtime settings", function()
    t.it("applies master volume and fullscreen through thin LÖVE adapters", function()
        local saved_audio = love.audio
        local saved_window = love.window
        local volume = nil
        local fullscreen = false
        love.audio = {
            setVolume = function(value)
                volume = value
            end,
        }
        love.window = {
            getFullscreen = function()
                return false
            end,
            setFullscreen = function(value)
                fullscreen = value
                return true
            end,
        }

        local value = settings.defaults()
        value.master_volume = 0.35
        value.fullscreen = true
        local ok, err = pcall(runtime_settings.apply, value)
        love.audio = saved_audio
        love.window = saved_window

        assert(ok, tostring(err))
        t.eq(volume, 0.35)
        t.eq(fullscreen, true)
    end)

    t.it("routes existing shake and bloom settings to reduced combat feedback", function()
        local value = settings.defaults()
        value.screen_shake = false
        value.bloom = false
        runtime_settings.apply(value)
        local diagnostics = combat_feedback.diagnostics(combat_feedback.new())
        t.is_true(diagnostics.reduced_motion)
        t.is_true(diagnostics.reduced_flash)
        runtime_settings.apply(settings.defaults())
    end)
end)

File: ts/packages/app/src/runtime_settings.spec.ts
// `runtime_settings.apply` forwards settings to a
// `CombatFeedbackDefaultsPort`. `@gc/presentation`'s `combat_feedback`
// (whose `diagnostics` would confirm the downstream effect on motion/flash)
// is not a declared dependency of this package, so the second case below
// asserts against a recording `CombatFeedbackDefaultsPort` double instead:
// it proves `runtime_settings` forwards the exact settings object to
// `configureDefaults`, which is this module's own contribution -- the
// downstream claim that `@gc/presentation` actually reduces motion/flash
// from it belongs to that package's own tests, once this package can
// depend on it.

import { describe, expect, it } from "vitest";
import { Audio, type CombatFeedbackPort } from "./audio.ts";
import { runtimeSettings, type RuntimeSettingsPorts } from "./runtime_settings.ts";
import { settings } from "./settings.ts";

const NOOP_COMBAT_FEEDBACK: CombatFeedbackPort = {
  link: () => ({ disposition: {} }),
  disposition: () => ({}),
};

function newPorts(overrides: Partial<RuntimeSettingsPorts> = {}): RuntimeSettingsPorts & {
  readonly bloom: { config: { enabled: boolean } };
  readonly configureDefaultsCalls: unknown[];
} {
  const configureDefaultsCalls: unknown[] = [];
  const effectsConfigureCalls: unknown[] = [];
  return {
    audio: new Audio(NOOP_COMBAT_FEEDBACK),
    combatFeedback: {
      configureDefaults: (value) => {
        configureDefaultsCalls.push(value);
      },
    },
    bloom: { config: { enabled: true } },
    effects: {
      configure: (value) => {
        effectsConfigureCalls.push(value);
      },
    },
    configureDefaultsCalls,
    ...overrides,
  };
}

describe("runtime settings", () => {
  it("applies master volume and fullscreen through thin browser-port adapters", () => {
    let volume: number | undefined;
    let fullscreen = false;
    const ports = newPorts({
      setMasterVolume: (value) => {
        volume = value;
      },
      window: {
        getFullscreen: () => false,
        setFullscreen: (value) => {
          fullscreen = value;
          return true;
        },
      },
    });

    const value = { ...settings.defaults(), master_volume: 0.35, fullscreen: true };
    runtimeSettings.apply(ports, value);

    expect(volume).toBe(0.35);
    expect(fullscreen).toBe(true);
  });

  it("routes existing shake and bloom settings to the combat feedback defaults port", () => {
    const ports = newPorts();
    const value = { ...settings.defaults(), screen_shake: false, bloom: false };
    runtimeSettings.apply(ports, value);

    expect(ports.configureDefaultsCalls).toEqual([value]);
    expect(ports.bloom.config.enabled).toBe(false);
  });
});
